ipld/retriever.go

```diff
@@ -0,0 +1,153 @@
+//  ---- ----
+// | 2  | 3  |
+//  ---- ----
+// Retriever randomly picks one of the data square quadrants and tries to request them one by one until it is able to
+// reconstruct the whole square.
+type Retriever struct {
+	dag   format.DAGService
+	codec rsmt2d.Codec
+}
+
+// NewRetriever creates a new instance of the Retriever over IPLD Service and rmst2d.Codec
+func NewRetriever(dag format.DAGService, codec rsmt2d.Codec) *Retriever {
+	return &Retriever{dag: dag, codec: codec}
+}
+
+// Retrieve retrieves all the data committed to DataAvailabilityHeader.
+// It aims to request from the network only 1/4 of the data and reconstructs other 3/4 parts using the rsmt2d.Codec.
+// It steadily tries to request other 3/4 data if 1/4 is not found within RetrieveQuadrantTimeout or unrecoverable.
+func (r *Retriever) Retrieve(ctx context.Context, dah *da.DataAvailabilityHeader) (*rsmt2d.ExtendedDataSquare, error) {
+	ses := r.newSession(ctx, dah)
+	for _, qs := range newQuadrants(dah) {
+		// TODO: ErrByzantineRow/Col
+		for _, q := range qs {
+			eds, err := ses.retrieve(ctx, q)
+			if err == nil {
+				return eds, nil
+			}
+		}
+		// retry quadrants until we have something or error
+	}
+
+	return nil, format.ErrNotFound
+}
+
+type retrieverSession struct {
+	dag   format.NodeGetter
+	adder *NmtNodeAdder
+
+	treeFn rsmt2d.TreeConstructorFn
+	codec  rsmt2d.Codec
+
+	dah    *da.DataAvailabilityHeader
+	square [][]byte
+}
+
+func (r *Retriever) newSession(ctx context.Context, dah *da.DataAvailabilityHeader) *retrieverSession {
+	size := len(dah.RowsRoots)
+	adder := NewNmtNodeAdder(ctx, format.NewBatch(ctx, r.dag, format.MaxSizeBatchOption(batchSize(size))))
+	return &retrieverSession{
+		dag:   merkledag.NewSession(ctx, r.dag),
+		adder: adder,
+		treeFn: func() rsmt2d.Tree {
+			tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(size)/2, nmt.NodeVisitor(adder.Visit))
+			return &tree
+		},
+		codec:  r.codec,
+		dah:    dah,
+		square: make([][]byte, size*size),
+	}
+}
+
+func (rs *retrieverSession) retrieve(ctx context.Context, q *quadrant) (*rsmt2d.ExtendedDataSquare, error) {
+	log.Debugw("retrieving data square", "data_hash", hex.EncodeToString(rs.dah.Hash()), "size", len(rs.square))
+	defer func() {
+		// all shares which were requested or repaired are written to disk with the commit
+		// we store *all*, so they are served to the network, including incorrectly committed data(BEFP case),
+		// so that network can check BEFP
+		if err := rs.adder.Commit(); err != nil {
+			log.Errorw("committing DAG", "err", err)
+		}
+	}()
+	// request quadrant and fill it into rs.square slice
+	// we don't care about the errors here, just need to request as much data as we can to be able to reconstruct below
+	rs.request(ctx, q)
+
+	// try repair
+	err := rsmt2d.RepairExtendedDataSquare(rs.dah.RowsRoots, rs.dah.ColumnRoots, rs.square, rs.codec, rs.treeFn)
+	if err != nil {
+		log.Errorw("not enough shares to reconstruct data square, retrying...", "err", err)
```

Comment:
```suggestion
		log.Errorw("not enough shares to reconstruct data square, requesting more...", "err", err)
```

Reply:
maybe make this a debug level log? How often will something like this pop up? In this case it's not really an "error" if we can still possibly reconstruct from other quadrants. We should only log at error level if reconstruction fails after *all* quadrants tried.

Reply:
Agree that this is not an error. I would do a warning instead. The reason is that retry is not a normal case. Usually, everything should be fine and without retries.

Reply:
Changed